app/controllers/regions_search_controller.rb

```diff
@@ -0,0 +1,49 @@
+class RegionsSearchController < AdminController
+  delegate :cache, to: Rails
+
+  def show
+    accessible_facility_regions = authorize { current_admin.accessible_facility_regions(:view_reports) }
+    cache_key = "#{current_admin.cache_key}/regions/index"
+    cache_version = "#{accessible_facility_regions.cache_key} / v2"
+    @accessible_regions = cache.fetch(cache_key, version: cache_version, expires_in: 7.days) {
+      accessible_facility_regions.each_with_object({}) { |facility, result|
+        ancestors = Hash[facility.cached_ancestors.map { |facility| [facility.region_type, facility] }]
+        org, state, district, block = ancestors.values_at("organization", "state", "district", "block")
+        result[org] ||= {}
+        result[org][state] ||= {}
+        result[org][state][district] ||= {}
+        result[org][state][district][block] ||= []
+        result[org][state][district][block] << facility
+      }
+    }
+    @query = params.permit(:query)[:query]
+    regex = /.*#{Regexp.escape(@query)}.*/i
+    results = search(@accessible_regions, regex)
+    json = results.sort_by(&:name).map { |region|
```

Comment:
Have you tried using this live? So far its very snappy and fast.  I'm fine using pg_search to get better scoring / ranking, but I think the simplest approach is better to get something live here.

Reply:
Oh, the other problem with a direct pg_search is that we then have to do authorization as a separate function on top of the search, because we can't easily ask for accessible regions until UserAccess has moved to Regions.

Reply:
I tried this out on sandbox, it was quite snappy, yep. Although that might be because we are caching everything on redis? I meant that the PG search query would be fast enough to not warrant any caching.

Is `accessible_regions` just the accessible facilities and all the parents? Would something like this work for now
```ruby
Region
  .where(id: accessible_facility_regions)
  .joins("inner join regions parent ON parent.path @> regions.path and parent.region_type != 'organization' and parent.region_type != 'root'")
```


Also I couldn't get it to run locally with a large data set, was running into this error
![image](https://user-images.githubusercontent.com/16774200/105220541-e242cb80-5b7d-11eb-8224-3cb9839a553a.png)

Reply:
We can (and should) make scopes out of these `ancestor` and related concepts. The regions table is just 2mb on production, so I don't know if we need an additional caching layer for stuff we can easily do in SQL via self-joins.